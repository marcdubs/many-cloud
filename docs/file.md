# File

- [Introduction](#introduction)
- [Constructors](#constructors)
  * [new File(id, connection)](#new-fileid-connection)
- [Methods](#methods)
  * [download_to(path)](#download_topath)
  * [get_checksum()](#get_checksum)

## Introduction

File is one of the three object classes that provides abstraction to what cloud drive you're dealing with. Providing similar functions for all of them.

File and Folder both import all of the functions inside of [item](item.md).

After you establish a connection to any integration as described in the [authentication](authentication.md) documentation, you can create a file like so:
```js
const File = require("many-cloud/lib/abstractions/file");

//Create a File object with an ID and a drive connection.
let file = new File("some-file-id", connection);
```

Along with all the functions of [item](item.md), file also has these functions:

## Constructors

### new File(id, connection)

Parameters:
| Name | Type | Description |
| -- | -- | -- |
| id | String | A file ID |
| connection | Object | A connection to an integration |

## Methods

### download_to(path)

Parameters:
| Name | Type | Description |
| -- | -- | -- |
| path | String | Path to destination |

Returns: Promise that resolves once file is downloaded

### get_checksum()

Parameters: None

Returns: Promise that resolves with a checksum (sha1 or md5, etc depending on what the cloud service decided on).
