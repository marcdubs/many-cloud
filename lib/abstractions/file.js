const BoxFile = require("../integrations/Box/abstractions/file");
const GoogleDriveFile = require("../integrations/GoogleDrive/abstractions/file");
const S3DriveFile = require("../integrations/S3/abstractions/file");

/**
 * File class
 * @constructor
 * @param {String} id - File id.
 * @param {Object} connection - A connection to an integration.
 */
function File(id, connection) {
  if (connection.type === "Box") {
    return new BoxFile(id, connection);
  } else if (connection.type === "GoogleDrive") {
    return new GoogleDriveFile(id, connection);
  } else if (connection.type === "S3") {
    return new S3DriveFile(id, connection);
  }
  throw new Error("Unrecognized connection type: " + connection.type);
}

module.exports = File;
