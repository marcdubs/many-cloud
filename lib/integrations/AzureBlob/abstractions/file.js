const Item = require("./item");

const fs = require("fs");
const tmp = require("tmp");
const path = require("path");
const md5File = require("md5-file");

class File extends Item {
  constructor(id, connection) {
    super(connection);

    this.type = "file";
    this.id = id;

    // Remove the first character if it is a slash
    if (this.id.charAt(0) === "/") {
      this.id = this.id.substring(1);
    }

    this.blockBlobClient = connection.containerClient.getBlockBlobClient(this.id);
  }

  async download_to(localPath) {
    if (fs.existsSync(localPath)) {
      fs.rmSync(localPath);
    }

    await this.blockBlobClient.downloadToFile(localPath);
  }

  async delete() {
    await this.blockBlobClient.delete();
  }

  /**
   * Not recommended to be heavily used for Azure, since this needs to download the file temporarily to get the checksum.
   */
  async get_checksum() {
    const tempDir = tmp.dirSync({
      unsafeCleanup: true,
    });

    let localFilePath = path.join(tempDir.name, await this.get_name());

    await this.download_to(localFilePath);

    const hash = md5File.sync(localFilePath);

    tempDir.removeCallback();

    return hash;
  }
}

module.exports = File;