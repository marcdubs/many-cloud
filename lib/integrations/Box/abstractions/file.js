const Item = require("./item");

/**
 * File class
 * @constructor
 * @param {String} id - File id.
 * @param {Object} connection - A connection to a Box integration.
 */
function File(id, connection) {
  this.id = id;
  this.type = "file";
  this.connection = connection;
}

File.prototype = Object.create(Item.prototype);
File.prototype.constructor = File;

/**
 * Downloads this file to specified location
 * @param {String} path - Path to destination.
 *
 * @return {Promise}
 */
File.prototype.download_to = function(path) {
  return new Promise(async (resolve, reject) => {
    try {
      await this.connection.download_file(this.id, path);
      resolve();
    } catch (err) {
      reject(err);
    }
  });
};

/**
 * Gets a checksum (sha1 or md5, etc depending on implementation)
 *
 * @return {Promise} Resolves with the checksum.
 */
File.prototype.get_checksum = function() {
  return new Promise(async (resolve, reject) => {
    if (!this.checksum) {
      try {
        await this.retrieve_info();
      } catch (err) {
        reject(err);
        return;
      }
    }
    resolve(this.checksum);
  });
};

module.exports = File;
