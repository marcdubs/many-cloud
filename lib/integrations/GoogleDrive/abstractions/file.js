const Item = require("./item");

/**
 * File class
 * @constructor
 * @param {String} id - File id.
 * @param {Object} connection - A connection to a Google Drive integration.
 */
function File(id, connection) {
  this.id = id;
  this.type = "file";
  this.connection = connection;
}

File.prototype = Object.create(Item.prototype);
File.prototype.constructor = File;

/**
 * Downloads this file to specified location
 * @param {String} path - Path to destination.
 *
 * @return {Promise}
 */
File.prototype.download_to = function(path) {
  return new Promise(async (resolve, reject) => {
    try {
      await this.connection.download_file(this.id, path);
      resolve();
    } catch (err) {
      reject(err);
    }
  });
};

module.exports = File;
