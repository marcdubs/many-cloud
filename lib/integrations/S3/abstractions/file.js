const Item = require("./item");

/**
 * File class
 * @constructor
 * @param {String} id - File id.
 * @param {Object} connection - A connection to a Google Drive integration.
 */
function File(id, connection) {
  this.id = id;
  this.type = "file";
  this.connection = connection;

  //For S3, the name of this file is the end of the key:
  //If the key is: some/path/to/textFile.txt then the file is called textFile.txt
  let split = id.split("/");
  this.name = split[split.length - 1];
}

File.prototype = Object.create(Item.prototype);
File.prototype.constructor = File;

/**
 * Downloads this file to specified location
 * @param {String} path - Path to destination.
 *
 * @return {Promise}
 */
File.prototype.download_to = function(path) {
  return new Promise(async (resolve, reject) => {
    try {
      await this.connection.download_file(this.id, path);
      resolve();
    } catch (err) {
      reject(err);
    }
  });
};

/**
 * Gets a checksum (sha1 or md5, etc depending on implementation)
 *
 * @return {Promise} Resolves with the checksum.
 */
File.prototype.get_checksum = function() {
  return new Promise(async (resolve, reject) => {
    if (!this.checksum) {
      try {
        await this.retrieve_info();
      } catch (err) {
        reject(err);
        return;
      }
    }
    resolve(this.checksum);
  });
};

module.exports = File;
