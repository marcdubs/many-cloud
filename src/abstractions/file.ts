import Item from "./item";

export default interface File extends Item {
  type: "file";
  download_to(localPath: string): Promise<void>;
  get_checksum(): Promise<string>;
}
