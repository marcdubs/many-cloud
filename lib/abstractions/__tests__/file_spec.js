const File = require("../file");

describe("file", () => {
  describe("constructor", () => {
    describe("when connection type is Box", () => {
      let file;
      beforeEach(() => {
        let connection = {
          type: "Box"
        };
        file = new File("1", connection);
      });

      it("returns a BoxFile", () => {
        expect(file).toBeInstanceOf(
          require("../../integrations/Box/abstractions/file.js")
        );
      });
    });

    describe("when connection type is GoogleDrive", () => {
      let file;
      beforeEach(() => {
        let connection = {
          type: "GoogleDrive"
        };
        file = new File("1", connection);
      });

      it("returns a GoogleDriveFile", () => {
        expect(file).toBeInstanceOf(
          require("../../integrations/GoogleDrive/abstractions/file.js")
        );
      });
    });

    describe("when connection type is S3", () => {
      let file;
      beforeEach(() => {
        let connection = {
          type: "S3"
        };
        file = new File("1", connection);
      });

      it("returns a S3DriveFile", () => {
        expect(file).toBeInstanceOf(
          require("../../integrations/S3/abstractions/file.js")
        );
      });
    });

    describe("when connection type is something else", () => {
      let connection;
      beforeEach(() => {
        connection = {
          type: "idk"
        };
      });

      it("throws an error", () => {
        expect(() => {
          new File("1", connection);
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
