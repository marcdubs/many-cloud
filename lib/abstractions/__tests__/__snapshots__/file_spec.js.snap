// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file constructor when connection type is something else throws an error 1`] = `"Unrecognized connection type: idk"`;
